### PR TITLE
[SPARK-12888][SQL] benchmark the new hash expression

### PR DESCRIPTION
**Pull request by cloud-fan**
Benchmark it on 4 different schemas, the result:

```
Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
Hash For simple:                   Avg Time(ms)    Avg Rate(M/s)  Relative Rate
-------------------------------------------------------------------------------
interpreted version                       31.47           266.54         1.00 X
codegen version                           64.52           130.01         0.49 X
```

```
Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
Hash For normal:                   Avg Time(ms)    Avg Rate(M/s)  Relative Rate
-------------------------------------------------------------------------------
interpreted version                     4068.11             0.26         1.00 X
codegen version                         1175.92             0.89         3.46 X
```

```
Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
Hash For array:                    Avg Time(ms)    Avg Rate(M/s)  Relative Rate
-------------------------------------------------------------------------------
interpreted version                     9276.70             0.06         1.00 X
codegen version                        14762.23             0.04         0.63 X
```

```
Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
Hash For map:                      Avg Time(ms)    Avg Rate(M/s)  Relative Rate
-------------------------------------------------------------------------------
interpreted version                    58869.79             0.01         1.00 X
codegen version                         9285.36             0.06         6.34 X
```

### CODE REVIEW

**Comment by cloud-fan** [comment]
The hash exprssion is slow on simple schema because the `UnsafeProjection` has some unnecessary opeations every round and will hurt the performance if each round is quite small(like simple schema), this can be fixed by https://github.com/apache/spark/pull/10809. For array type, still investigate.

cc @nongli

**Comment by cloud-fan** [comment]
retest this please

**Comment by nongli** [comment]
I'm confused by the conclusino. We don't use UnsafeProjection anymore right?

**Comment by cloud-fan** [comment]
We need `UnsafeProjection` to execute the codegen version of hash expression.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/HashBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      .add("string", StringType)
+      .add("binary", BinaryType)
+      .add("date", DateType)
+      .add("timestamp", TimestampType)
+    test("normal", normal, 128)
+
+    val arrayOfInt = ArrayType(IntegerType)
+    val array = new StructType()
+      .add("array", arrayOfInt)
+      .add("arrayOfArray", ArrayType(arrayOfInt))
```
Comment (nongli) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/HashBenchmark.scala]: Can you include the results as comments for each test() case?

**Comment by nongli** [comment]
LGTM

**Comment by cloud-fan** [comment]
we should run this benchmark again after https://github.com/apache/spark/pull/10809 is merged.

**Comment by rxin** [comment]
I'm going to merge this. Please submit an update once #10809 is merged and put the result in the test case as comments.

**Pull request closed without merging**